File: apps/looperd/src/infra/spec-pr.ts
import type { GitHubPullRequestDetail } from "./github";

export const SPEC_REVIEWING_LABEL = "looper:spec-reviewing";
export const SPEC_READY_LABEL = "looper:spec-ready";
export const NEEDS_HUMAN_LABEL = "looper:needs-human";

export type PullRequestPhase = "spec" | "implementation";

export function normalizeLabel(label: string): string {
  return label.trim().toLowerCase();
}

export function hasLabel(labels: string[] | undefined, label: string): boolean {
  const normalizedTarget = normalizeLabel(label);
  return (labels ?? []).some(
    (candidate) => normalizeLabel(candidate) === normalizedTarget,
  );
}

export function resolvePullRequestPhase(input: {
  labels?: string[];
}): PullRequestPhase {
  return hasLabel(input.labels, SPEC_REVIEWING_LABEL)
    ? "spec"
    : "implementation";
}

export function parseSpecPathFromPullRequestBody(
  body: string | null | undefined,
): string | null {
  if (!body) {
    return null;
  }

  const match = /^Spec:\s*(.+)$/im.exec(body);
  return match?.[1]?.trim() || null;
}

export function countUnresolvedReviewThreads(comments: unknown[]): number {
  return comments.filter((comment) => {
    if (!comment || typeof comment !== "object") {
      return false;
    }

    const record = comment as Record<string, unknown>;
    const state = record.state;
    if (typeof state === "string") {
      return state.toUpperCase() !== "RESOLVED";
    }

    return record.isResolved !== true;
  }).length;
}

export function isSpecReviewClean(
  detail: Pick<GitHubPullRequestDetail, "comments" | "reviewDecision">,
): boolean {
  return (
    countUnresolvedReviewThreads(detail.comments) === 0 &&
    detail.reviewDecision?.toUpperCase() !== "CHANGES_REQUESTED"
  );
}

File: internal/infra/specpr/specpr.go
package specpr

import (
	"regexp"
	"strings"
)

const (
	ReviewingLabel  = "looper:spec-reviewing"
	ReadyLabel      = "looper:spec-ready"
	NeedsHumanLabel = "looper:needs-human"
)

type PullRequestPhase string

const (
	PhaseSpec           PullRequestPhase = "spec"
	PhaseImplementation PullRequestPhase = "implementation"
)

var pathPattern = regexp.MustCompile(`(?mi)^Spec:\s*(.+)$`)

func NormalizeLabel(label string) string {
	return strings.ToLower(strings.TrimSpace(label))
}

func HasLabel(labels []string, target string) bool {
	normalizedTarget := NormalizeLabel(target)
	for _, label := range labels {
		if NormalizeLabel(label) == normalizedTarget {
			return true
		}
	}
	return false
}

func ResolvePullRequestPhase(labels []string) PullRequestPhase {
	if HasLabel(labels, ReviewingLabel) {
		return PhaseSpec
	}
	return PhaseImplementation
}

func ParseSpecPathFromPullRequestBody(body string) string {
	matches := pathPattern.FindStringSubmatch(body)
	if len(matches) < 2 {
		return ""
	}
	return strings.TrimSpace(matches[1])
}

func CountUnresolvedReviewThreads(comments []map[string]any) int {
	count := 0
	for _, comment := range comments {
		state, _ := comment["state"].(string)
		if state != "" {
			if !strings.EqualFold(state, "RESOLVED") {
				count++
			}
			continue
		}
		isResolved, _ := comment["isResolved"].(bool)
		if !isResolved {
			count++
		}
	}
	return count
}

func IsReviewClean(reviewDecision string, comments []map[string]any) bool {
	return CountUnresolvedReviewThreads(comments) == 0 && !strings.EqualFold(reviewDecision, "CHANGES_REQUESTED")
}
